Change dockerfile from using Node 19 to match dev environment

rubys opened the pull request:
### Motivation / Background

This Pull Request has been created because deploying with a different node version than the application was tested with locally may not work as intended (or even at all).

### Detail

This Pull Request changes the generated dockerfile to install the locally installed versions of node and yarn.  It also:

  * decouples the installation of libvips from node
  * adds installation of the node_modules by running yarn

### Additional information

This pull request is part one of many intended to add features from fly.io's generated Dockerfiles that are of general interest to Rails.  For the overall plan, see:
https://community.fly.io/t/preparations-for-rails-7-1/9512

I'm making these changes in stages so that I can get early feedback and in so doing improve future pull requests.

This changes was tested by generating an new application with esbuild and deploying it.

attn: @dhh

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

rubys committed:
Change dockerfile from using Node 19 to match dev environment

This is accomplished via installing volta: https://volta.sh/.

Also:
  * decouple the installation of libvips from node
  * add installation of the node_modules by running yarn

This pull request is part one of many intended to add features
from fly.io's generated Dockerfiles that are of general interest
to Rails.  For the overall plan, see:
https://community.fly.io/t/preparations-for-rails-7-1/9512

I'm making these changes in stages so that I can get early feedback
and in so doing improve future pull requests.

This changes was tested by generating an new application with esbuild
and deploying it.

attn: @ddh

rubys commented:
It is not clear to me how to test a generated dockerfile in a CI environment.  And it has been a while since I have contributed to Rails, but [Running tests for a specific directory](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#for-a-specific-directory) doesn't work for me:

```
rubys@rubym1p railties % bin/test generators
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/date_time_parser.rb:837: warning: statement not reached
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/date_time_parser.rb:691: warning: assigned but unused variable - testEof
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32457: warning: statement not reached
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32618: warning: statement not reached
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32654: warning: statement not reached
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32743: warning: statement not reached
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32759: warning: statement not reached
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32825: warning: statement not reached
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32866: warning: statement not reached
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32891: warning: statement not reached
/opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:31987: warning: assigned but unused variable - testEof
/opt/homebrew/lib/ruby/gems/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require': cannot load such file -- test_helper (LoadError)
	from /opt/homebrew/lib/ruby/gems/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
	from /Users/rubys/git/rails/railties/test/isolation/assets/node_modules/@rails/actioncable/test/channel/base_test.rb:3:in `<top (required)>'
	from /opt/homebrew/lib/ruby/gems/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
	from /opt/homebrew/lib/ruby/gems/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
	from /Users/rubys/git/rails/railties/lib/rails/test_unit/runner.rb:48:in `block in load_tests'
	from /Users/rubys/git/rails/railties/lib/rails/test_unit/runner.rb:48:in `each'
	from /Users/rubys/git/rails/railties/lib/rails/test_unit/runner.rb:48:in `load_tests'
	from /Users/rubys/git/rails/railties/lib/rails/test_unit/runner.rb:41:in `run'
	from /Users/rubys/git/rails/tools/test.rb:18:in `<top (required)>'
	from bin/test:5:in `require_relative'
	from bin/test:5:in `<main>'
```

rubys committed:
switch to double quotes to placate Rubocop

palkan commented on railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:
```diff
@@ -3,16 +3,16 @@ ARG RUBY_VERSION=<%= Gem.ruby_version %>
+# Install JavaScript dependencies
+ARG NODE_VERSION=<%= dockerfile_node_version %>
+ARG YARN_VERSION=<%= dockerfile_yarn_version %>
+RUN curl https://get.volta.sh | bash
```
I'm a bit concerned of using third-party vendors; maybe, we can use official sources? Here is how we do this in [ruby-on-whales](https://github.com/evilmartians/ruby-on-whales/blob/c83cd5f74b83bfc6275179c274619f3f1ed7b7d3/example/.dockerdev/Dockerfile#L38-L47):

```dockerfile
ARG NODE_MAJOR
RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
  --mount=type=cache,target=/var/lib/apt,sharing=locked \
  --mount=type=tmpfs,target=/var/log \
  curl -sL https://deb.nodesource.com/setup_$NODE_MAJOR.x | bash - && \
  DEBIAN_FRONTEND=noninteractive apt-get install -yq --no-install-recommends \
    nodejs

ARG YARN_VERSION
RUN npm install -g yarn@$YARN_VERSION
```

(Ignore `--mount=type=cache`, it worths a separate PR).

rubys replied:
I would not be happy with that change.

Scenario: At the beginning of the month, after running rigorous regression tests I deploy an application.  Gemfile.lock says stimulus-rails version 1.2.1 so that's the version that is deployed.  yarn.lock says @hotwired/turbo version 7.2.4 so that is the version that is deployed.  At the time of the release Node 19.2.0 is the latest version of node, so that is what is deployed.

Mid month there is a spike in traffic so I deploy a second server.  It gets stimulus-rails 1.2.1, @hotwired/turbo 7.2.4, and.... node 19.3.0.

I realize that dockerfiles aren't completely reproducible.  I may, for example, get a different version of libvips the next time I deploy, but that is a bit more isolated than the node.js runtime upon which a large number of packages depend.

palkan replied:
Got it (but still don't think upgrading to a new **minor version** of Node could be too harmful; libvips or any other system lib could be).

Another approach (which I've been using in production) is to copy Node files from an official image:

```ruby
FROM node:16-alpine3.16 AS node
FROM ruby:3.1.3-alpine3.16 as main

COPY --from=node /usr/lib /usr/lib
COPY --from=node /usr/local/share /usr/local/share
COPY --from=node /usr/local/lib /usr/local/lib
COPY --from=node /usr/local/include /usr/local/include
COPY --from=node /usr/local/bin /usr/local/bin
COPY --from=node /opt /opt
```

rubys replied:
I've seen breaking changes in openssh causing upgrade problems for node.js users.

I see you copying from alpine to alpine which should be fine.  Ruby's latest image is based on bullseye.  Some of the previous images are based on buster.  Aligning the node and ruby images would be a bit of a maintenance nightmare.

palkan commented:
> This changes was tested by generating an new application with esbuild and deploying it.

I wonder if we need some automated testing for Dockerfiles 🤔 At least, smoke tests (generate new app and make sure a Docker image could be built successfully).

skipkayhil commented:
> It is not clear to me how to test a generated dockerfile in a CI environment.  And it has been a while since I have contributed to Rails, but [Running tests for a specific directory](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#for-a-specific-directory) doesn't work for me:
> 
> 
> 
> ```
> 
> rubys@rubym1p railties % bin/test generators
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/date_time_parser.rb:837: warning: statement not reached
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/date_time_parser.rb:691: warning: assigned but unused variable - testEof
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32457: warning: statement not reached
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32618: warning: statement not reached
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32654: warning: statement not reached
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32743: warning: statement not reached
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32759: warning: statement not reached
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32825: warning: statement not reached
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32866: warning: statement not reached
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:32891: warning: statement not reached
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/mail-2.8.0/lib/mail/parsers/address_lists_parser.rb:31987: warning: assigned but unused variable - testEof
> 
> /opt/homebrew/lib/ruby/gems/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require': cannot load such file -- test_helper (LoadError)
> 
> 	from /opt/homebrew/lib/ruby/gems/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
> 
> 	from /Users/rubys/git/rails/railties/test/isolation/assets/node_modules/@rails/actioncable/test/channel/base_test.rb:3:in `<top (required)>'
> 
> 	from /opt/homebrew/lib/ruby/gems/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
> 
> 	from /opt/homebrew/lib/ruby/gems/3.1.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
> 
> 	from /Users/rubys/git/rails/railties/lib/rails/test_unit/runner.rb:48:in `block in load_tests'
> 
> 	from /Users/rubys/git/rails/railties/lib/rails/test_unit/runner.rb:48:in `each'
> 
> 	from /Users/rubys/git/rails/railties/lib/rails/test_unit/runner.rb:48:in `load_tests'
> 
> 	from /Users/rubys/git/rails/railties/lib/rails/test_unit/runner.rb:41:in `run'
> 
> 	from /Users/rubys/git/rails/tools/test.rb:18:in `<top (required)>'
> 
> 	from bin/test:5:in `require_relative'
> 
> 	from bin/test:5:in `<main>'
> 
> ```

I've seen similar issues to this. The problem is the require is globbing _any_ file inside the test directory, and it's finding files in node_modules (which it shouldn't). I've been fixing it by removing that node_modules folder (since it's created somewhere during tests)

jonathanhefner commented:
> It is not clear to me how to test a generated dockerfile in a CI environment.

If you just want to test the generated Dockerfile, you can run:

```console
$ cd railties
$ bin/test test/generators/app_generator_test.rb -n /docker/
```

`/docker/` is a regex that matches the test method names.

We only have [a few assertions](https://github.com/rails/rails/blob/28386bdc0fedecc275c3f1bf0a271beafb0a8ba5/railties/test/generators/app_generator_test.rb#L1012-L1026) currently, but as the Dockerfile becomes more complex, it would be good to add more.

rubys committed:
add some Dockerfile tests

rubys commented:
Tests passed before my last push.  Current test failures appear to be unrelated to the change.

rubys commented:
Further investigation:

  * build-essential is required in order to gem install rails (and therefore to bundle install any rails application).  Multi-stage builds can be used to prevent the additional overhead of compilers and libraries from being present in the final deployed image. 
  * It is important to match bundler versions too when doing a bundle install

Those changes can go into a subsequent pull request.

It is also not clear to me whether refining a new feature merits additional changelog entries.  Any guidance that can be provided here would be appreciated.

palkan commented:
> It is important to match bundler versions too when doing a bundle install

It’s not necessarily with recent Bundler versions; now Bundler installs the version from the lockfile automatically.

rubys committed:
generate a .node-version file if using_node?

rubys commented:
@dhh I added a .node-version file.  More needed to be done?  Not interested?

dhh commented:
I like this .node-version approach, but it seems like we need to solve this entire question outside of the context of the Dockerfile. That is, we'd need to have Volta involved in setting up the app with bin/setup. Possibly bailing if things aren't matching, like we do with .ruby-version.

rubys commented:
@dhh I'm not following.  To further the discussion, let me explain my perspective, and perhaps then you can tell me what I am missing.

There are multiple ruby version managers.  rvm, rbenv, and chruby are the top three.  Rails doesn't pick one, or even require you to use one.  It merely notes the current ruby when you create an application and produces a `.ruby-version` file using a template.  If you don't use a ruby version manager, this file is harmless, but doesn't prevent you from running the wrong version.  If you do use a version manager, this file can automatically switch you to the right version of Ruby.

There are multiple node version managers.  nvm, n, and volta are three examples, but there are plenty more.  Rails doesn't pick one, or even require you to use one.  It merely notes the current node when you create an application and produces a `.node-version` file using a template.  If you don't use a node version manager, this file is harmless, but doesn't prevent you from running the wrong version.  If you do use a version manager, this file can automatically switch you to the right version of Node.

A list of node tools that respect the `.node-version` file: https://github.com/shadowspawn/node-version-usage#node-version-file-usage

P.S.  This is just the first of many changes I am offering to make.  See: https://community.fly.io/t/rails-cookbooks/9918 for more details.  My preference is not for a single uber commit that solves all the potential problems, but a steady stream of smaller changes.  In that vein, let me know what is the minimum needed to declare victory and to get this pull request to land, and also let me know what change you would like to see explored next.

dhh commented on railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:
```diff
@@ -33,6 +33,12 @@ ENV RAILS_LOG_TO_STDOUT="1" \
+<% if using_node? -%>
+# install node modules
+COPY yarn.* ./
+RUN yarn install
```
This is already happening as part of the `assets:precompile`. Are you separating it out as a layer optimization?

rubys replied:
I removed this for now.  Yes, a `yarn install` without any caching whatsoever is rather time consuming, and not exactly an ideal user experience.  And there is a range of options from "I don't care how long it takes, I just want my Dockerfile to be gorgeous"  to "I don't care how complicated my Dockerfile is, I just want deploys to be fast".

I explore these options on my Rails cookbooks page.  Picking one approach now is premature.  Let's discuss: I don't care whether the discussion happens in git issues, Rails Discourse, Fly's community forum or elsewhere.  I'm flexible.

dhh replied:
`assets:precompile` is run as part of image creation, so I believe it's the same experience. It's all this done at the image stage. Don't think there's a cache difference?

dhh commented:
Didn't realize .node-version was already widely adopted. That does change the picture for me. Thought we were basically just modelling it on .ruby-version. So +1 on that from here.

What's the story with yarn version?

dhh commented:
Also, very much agree with taking this as piecemeal rather than big bang. Much easier to reason about.

rubys committed:
remove this for now

dhh commented on railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:
```diff
@@ -3,16 +3,16 @@ ARG RUBY_VERSION=<%= Gem.ruby_version %>
+RUN curl https://get.volta.sh | bash
+ENV VOLTA_HOME /root/.volta
+ENV PATH $VOLTA_HOME/bin:/usr/local/bin:$PATH
+RUN volta install node@$NODE_VERSION yarn@$YARN_VERSION
```
Think we should be combining some of these ARG, RUN, and ENV commands to reduce the layer overhead.

rubys committed:
reduce the number of layers

also adopt some of the style conventions already in use in the
Dockerfile with respect to quoting and spacing.

rubys commented:
> What's the story with yarn version?

I've yet to use it myself, but yarn does have a story.  See: https://yarnpkg.com/cli/set/version .  I want to experiment with it myself a bit before I propose a patch.  Since it downloads yarn, I want to do things like build on an m1 mac and deploy to a 686 linux server; and I also want to examine the artifacts to see which (if any) should be added to .gitignore and which should be committed and pushed to git.

dhh commented:
Think this otherwise looks good to go, yeah?

rubys commented:
My recommendation is to merge this pull request at which point I'll begin on the next one.  Whether yarn version is top of the list or sometime later matters not to me.  I'm confident that yarn version is a solvable problem so a part of me is yearning to explore other issues first.  Let me know.

dhh commented:
Great. Let's add a CHANGELOG note about generating the .node-version file by default now. Don't need to mention anything about the Dockerfile changes, since that's already covered. But should note that there's a new file for node apps on fresh generation.

rubys commented:
Agreed that's notable.  I'm concerned about merge conflicts.  Should I rebase and squash?  Or just rebase?  Or not worry about it?

dhh commented:
No worries. You can just add it, and I'll be squashing and merging right after 👍

rubys committed:
add changelog entry

dhh commented:
Thanks, Sam!

dhh commented:
So seems like we have to bail on this and revert, since it's not working on arm? The Dockerfile must work on arm out of the box.

dhh commented:
Opened https://github.com/rails/rails/issues/47441 to track rolling this back.

zzak commented:
Seems like adding `.node-version` and `.yarn-version` as discussed here could be separate but still valuable PRs?

dhh commented:
Not unless we run a multi-arch strategy for Docker that includes them or they'll give a false sense of version fixedness. Either way, need to restore main to working order first. Then we can go back to discussing upgrades.
